scripts/check.sh

```diff
@@ -0,0 +1,153 @@
+can review them with git diff. Pass --commit to automatically commit any changes.
+
+Pass an alternate revision to use as the basis for checking changes.
+
+OPTIONS:
+
+  --allow-dirty
+    By default, check.sh requires a clean working tree to keep any generated
+    changes separate from logical changes.
+
+  --commit
+    Commit any auto-generated changes.
+
+  <revision>
+    Specifies a starting revision other than the default of master.
+
+
+EXAMPLES:
+
+  check.sh
+    Runs automated checks and formatters on all changed files since master.
+    Check for changes with git diff.
+
+  check.sh --commit
+    Runs automated checks and formatters on all changed files since master and
+    commits the results.
+
+  check.sh --commit HEAD
+    Runs automated checks and formatters on all changed files since the last
+    commit.
+
+  check.sh --allow-dirty HEAD
+    Runs automated checks and formatters on all changed files since the last
+    commit and intermingles the changes with any pending changes. Useful for
+    interactive use from an editor.
+
+EOF
+}
+
+set -euo pipefail
+unset CDPATH
+
+# Change to the top-directory of the working tree
+top_dir=$(git rev-parse --show-toplevel)
+cd "$top_dir"
+
+allow_dirty=false
+commit=false
+start="master"
+
+while [[ $# -gt 0 ]]; do
+  case "$1" in
+    --)
+      # Do nothing: explicitly allow this, but ignore it
+      ;;
+
+    -h | --help)
+      usage
+      exit 1
+      ;;
+
+    --allow-dirty)
+      allow_dirty=true
+      ;;
+
+    --commit)
+      commit=true
+      ;;
+
+    *)
+      if git rev-parse "$1" >& /dev/null; then
+        start="$1"
+        break
+      fi
+      ;;
+  esac
+  shift
+done
+
+if [[ $allow_dirty == true && $commit == true ]]; then
+  echo "--allow_dirty and --commit are mutually exclusive"
+  exit 1
+fi
+
+if ! git diff-index --quiet HEAD --; then
+  if [[ $allow_dirty != true ]]; then
+    echo "You have local changes that could be overwritten by this script."
+    echo "Please commit your changes first or pass --allow-dirty."
+    exit 2
+  fi
+fi
+
+# Record actual start
+start_sha=$(git rev-parse "$start")
+
+# Restyle and commit any changes
+scripts/style.sh "$start_sha" || :
+if ! git diff --quiet; then
+  if [[ $commit == true ]]; then
+    echo "Style generated changes"
+    git commit -a --fixup=HEAD
+  fi
+fi
+
+# If there are changes to the Firestore project, ensure they're ordered
+# correctly to minimize conflicts.
+if ! git diff --quiet --name-only --diff-filter=ACMR -- \
+    Firestore/Example/Firestore.xcodeproj; then
+
+  scripts/sync_project.rb
+  if ! git diff --quiet; then
+    if [[ $commit == true ]]; then
+      echo "Sync Xcode project"
+      git commit -a --fixup=HEAD
+    fi
+  fi
+fi
+
+# Check lint errors
+(
+  scripts/lint.sh "$start_sha"
+  scripts/check_copyright.sh
+  scripts/check_no_module_imports.sh
+  scripts/check_test_inclusion.py
+  scripts/check_whitespace.sh
+) 2>&1 | sed "s,^\\([A-Za-z]*/\\),$top_dir/\\1,"
```

Comment:
Your sed script is adding top_dir back into the path (which is kinda annoying if you're already in the topdir... but I can see why you'd otherwise want it. Especially if you're looking to hook this into an IDE.) It'll break on anything with an initial component that doesn't match A-Za-z though. For instance, if we hooked up pod_lib_lint.sh and pointed it at FirebaseFirestore.podspec, it would seem like the sed script would fail. (Due to the `.` but also the lack of `/`)

I think, ideally, all of the scripts would run in whatever directory they're run in (rather than cd'ing into topdir) and then output the names of the matching files relative to the current directory (or always as an absolute path, though I'd find that noisy). That way, you wouldn't have to attempt to write a sed script to handle all cases. Seems like that might be more longer term though.

No action required.

Reply:
The issue I'm trying to fix here is that CLion only makes file:line clickable if the path is absolute, but these checks and linters are terrible about having well-formed output.

I've pulled this post-processing into a local wrapper.

Reply:
Feel free to keep it in if it's useful. (It seems like most ides are going to want absolute paths and although the sed script won't handle all cases, it seems like it should handle all current ones.)

Another random suggestion (entirely separate from this review); does clion behave better if you change it's working directory? (Easiest way to do that being to cd into topdir and start it from the shell.)

Reply:
It's cumbersome to optionally make paths absolute because the linters don't do a good job of having regular output. I may re-add this if I can come up with a better way to handle this universally without checking for `^Firestore` (or extending that to a regex including all top-level paths).